### PR TITLE
feat(components): Combobox Header

### PR DESCRIPTION
**Pull request by taylorvnoj**
## Motivations

For `Combobox` multi-select, we want an option to either `Select all` or `Clear` multiple selections, as well as display a counter for the number of options selected.

Screenshot 2023-10-16 at 10 11 19 AM

### Added

There is now a `ComboboxHeader` when `multiSelect = true`.

The label should:
- read `Select` || `Select {subjectNoun}` when no options are selected
- read `{selectedCount} selected` when one or more options is selected

The action button should:
- read `Select all` when no options are selected
- read `Clear` when one or more options is selected

As soon as a selection has been made, the action button will always display `Clear`. This means that you can `Select all` in a filtered state - only if you've made no selections prior to filtering.  For example:

We've filtered to find both Baggins' and can `Select all`:

Screenshot 2023-10-16 at 10 44 37 AM

However if I've previously selected Pippin Took, and then filter for the Baggins', I will need to manually select Bilbo and Frodo.  

Screenshot 2023-10-16 at 10 45 41 AM

This is intended behaviour as filtering should narrow down options to a very select few.

### Changed

`setInternalSelected` is now being used within `optionsSelectionHandler`.  
We need to make sure this new method of option selection within the `ComboboxHeader` will still work when `onSelect` or `onClose` are passed into `ComboboxContent`.

Because that logic (using either `props.onSelect` or `setInternalState`) was also being used in `handleSelection` we decided to extract that ternary and put it in `useComboboxContent` - named `optionsSelectionHandler`.

## Testing

`Trigger Chip` example is no multi-select with a subjectNoun.
`Multi Select` example of course will also display the `ComboboxHeader`

Make sure the `Select all` and `Clear` functionality works, and ensure the counter is displaying the correct amount of options selected.

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

Random photo of Atlantis

### CODE REVIEW

File: packages/components/src/Combobox/ComboboxProvider.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/ComboboxProvider.tsx:10]: do we need this?

I'm not seeing it being passed in nor exposed in the `value`

Comment (taylorvnoj) [reply packages/components/src/Combobox/ComboboxProvider.tsx:10]: Oops! We do not! Thank you!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx:54]: I'm assuming this is temporary code

a note for later, we'll need to make this look at the `selectedOptions` wrapper that will either be the external state passed in, or the internal state if we have the `onClose` callback

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.css
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.css:3]: 💅 tokens pls and thank u

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx:54]: Correct - just wanted something to actually change the label within the Header.  Which also means the Header component itself has temporary code based on just 1 selection.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.css
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.css:3]: lol let it be known that I know this - but was quickly putting px's instead of tokens. Still......shame!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx:93]: nice I like this approach where we figure out the logic of the count out here, and `ComboboxHeader` gets to be very naive and only cares about a number

I'm not sure when exactly we'd reuse it, but if we did it makes this component not coupled to our Combobox implementation

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.test.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.test.tsx:1]: for these `ComboboxHeader` tests, since we are unit testing `ComboboxHeader` which only has props, and never interacts with the Provider - we can actually completely omit that and just do

```
render(<Comboboxheader selectedCount={someNumber}/>)`
```

and then delete the `MockComboboxProvider` function definition in this file entirely

on that, if for some reason we did end up in a scenario where it was necessary to have the `MockComboboxProvider` wrapping these tests, I'd recommend exporting the existing definition and reusing that everywhere so we don't have multiple definitions of it that we need to keep in sync as changes happen!

**Comment by chris-at-jobber** [comment]
This looks really good so far! Seeing it live I'd like to adjust the spacing on the right of the header to have the action visually align with the selection icon ([updated Figma](https://www.figma.com/file/gQNDrpJYRFAwfbP3jRO0an/DSC-%5B-Combobox-%5D?type=design&node-id=1514%3A142971&mode=design&t=ohXz3qq9Wq3mRsQE-1) for reference). 

Playing with it in the code I think the specific adjustment I'm looking for is reducing the `padding-right` to `space-small`
image

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx:18]: assuming this is also temp

once it's real, we can probably make a single var of `const hasSelected = props.selectedCount > 0`

and then use that for the actionText and this bit of logic to show the `N selected` or the label

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.test.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.test.tsx:1]: Ahh ok that makes sense! 👍

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx:18]: Exactly, also temp.  Okay I love that and will simplify this nicely

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.tsx:108]: just a heads up, this won't work for the case where we were given `onSelect`

the internal state is only used for `onClose` where the passed in state of `selected` can diverge from the internal state

for `onSelect` we'll need to use the function we were given. you can see an example of how this work in `handleSelection` where we either use the `props.onSelect` or the `setInternalState` depending on if it's an `onClose` or not

**Comment by ZakaryH** [comment]
this is looking very nice!

there's a few cases I think would be valuable to have tests for because for one thing they are somewhat obscure flows, so knowing we have coverage on them will be extra valuable compared to flows we'll often see. secondly, the behavior is a bit debatable - so having the tests is a light form of self documentation that lets people know we are aware of the behavior and it is intended

we kinda have these tested already in the scenario where `filteredOptions` just is `options` but we don't have coverage on the scenarios where they diverge!

**Contextual Select All**

- have options A, B, C, AA
- type A in the search
- hit select all
- the selection handler should only have been called A and AA

**Not Contextual Clear**
- have options A, B, C, AA
- either have something already selected, or select it in the test
- type in A
- hit clear
- all selections should be cleared regardless of the search term

then there's on case I'm on the fence about where we pass it `onClose` instead. in this case, we would expect that the handler is not called when we press Select All, but they should be visually selected

this could be done as a pretty thorough integration test.... or we might be able to simply assert that `aria-selected` is true under those conditions

**Comment by ZakaryH** [comment]
couple things here

1) the width of the listbox changes because the text is changing, is that OK?
2) we can't guarantee that the `subjectNoun` won't run right up against the Select All button, is that OK?
Kapture 2023-10-16 at 12 55 28


pretty much the same thing, but taken to an extreme - with an absurdly long subjectNoun, it wraps and makes that same problem even more pronounced. I'm tempted to say this is fine. if you give us a really long word, that's just what happens
Kapture 2023-10-16 at 13 03 25

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx:16]: we've got a few bits of logic here that can be thematically grouped around labels

thinking ahead, as a courtesy to future us (or other developers) let's try to extract this out to a function so that we have as much room as possible before hitting the ol' `Function ComboboxHeader has too many statements (13)` at which point someone will be forced to refactor things

what do you think about this?

```
  const hasSelected = props.selectedCount > 0;
  const actionLabel = hasSelected ? "Clear" : "Select all";
  const label = getLabel(hasSelected, props.selectedCount, props.subjectNoun);
  const handleSelectAll = hasSelected ? props.onClearAll : props.onSelectAll;
```

where `getLabel` is a function defined outside the body of the `ComboboxHeader` to avoid re-making it

```
function getLabel(
  hasSelected: boolean,
  count: number,
  subjectNoun?: string,
): string {
  if (hasSelected) {
    return `${count} selected`;
  }

  if (subjectNoun) {
    return `Select ${subjectNoun}`;
  }

  return "Select";
}
```

Comment (ZakaryH) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx:16]: admittedly this is a bit heavy on ternaries, but I am of the opinion these are pretty easy to understand - so they're an appropriate use that keeps the code lean and doesn't sacrifice readability

**Comment by taylorvnoj** [comment]
> the width of the listbox changes because the text is changing, is that OK?

Although it's a bit choppier in this instance, the same thing happens when entering a long search (width grows).  I defer to @chris-at-jobber on if that's ok or not.

> we can't guarantee that the subjectNoun won't run right up against the Select All button, is that OK?

ahhh...I don't think that's ok. I can probably set something so there's always a 16px space between the end of the `subjectNoun` and the beginning of the header button - just like we have when there is a long search term

Screenshot 2023-10-16 at 4 32 30 PM


> with an absurdly long subjectNoun, it wraps

I wonder if this is the lesser or two evils where the other option is that Combobox can expand an entire page, hypothetically.

**Review by ZakaryH — CHANGES_REQUESTED** [review]
a couple questions around when subjectNoun is a long word and if we wanna adjust anything there. could be as simple as some padding, or maybe it's fine! this one is probably more of a @chris-at-jobber question

a couple tests that I think would be valuable to add

then a small suggestion for getting our lines of code down inside `ComboboxHeader` to avoid future refactors for as long as possible!

**Comment by chris-at-jobber** [comment]
I think these are the answers you're looking for...

I think it's outside the scope of this PR, but I don't love the container width dynamically re-sizing based on the length of its' children (ie after first render). For the header label, that should be static right? But can we constrain the search input to not grow/shrink the container as the user types? That one really bothers me.

If we can keep a `space--smaller` gap between the label and the action, I think that's sufficient.

I prefer the label wrapping to the container re-sizing.

**Comment by ZakaryH** [comment]
>I wonder if this is the lesser or two evils where the other option is that Combobox can expand an entire page, hypothetically.

we've got a few options here I feel. we could:
- truncate since the `subjectNoun` should be legible in the search, one could likely fill in the blanks at what the word is having already seen the full version above
- not show the subjectNoun in this section at all :o and keep it "Select` or maybe `Select Option` or `Select Options` for multiselect

**Comment by taylorvnoj** [comment]
@chris-at-jobber Width of the header label being static: yes and no?  In one of @ZakaryH  gifs, if you have a longer `subjectNoun`, once you 'Select all' the content box shrinks again because the label has shrunk down to say "1 selected". So the container width does re-size.

We can either :
- be ok with this possibly rare use case (not saying it's not important, just wondering how often `subjectNoun` will be 1. actually assigned and 2. a very long word)
or
- wrap the header label beyond 9/10 characters for the `subjectNoun` so that there is never any resizing
or
- expand the width of the content to give us a little more wiggle room


Either way I'm going to make a separate ticket to stop the grow/shrink of the container as the user types in the search input.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx]: Great call out, thank you Zak! I fixed this here 82d4f425d907227f5b358cd59976fb8afc98563c
I was previously heavy on ternaries in this file so went the opposite way.  Because these ternaries are 1 line I'm totally fine with that for the sake of keeping things lean

**Comment by taylorvnoj** [comment]
@chris-at-jobber `space-smaller` gap added here: 8f9eff3da919358c7d92d5a51db595cb23cf3b41

@ZakaryH let me know what you think about the tests I added here: 3a35cd2fcd3b2930f1d733ae012916e46784d151 - for the `onClear` I would agree I'm on the fence too.  Although I understand that we want it to be clear this Header functionality will be the same regardless of which prop is used, we are documenting that idea in other tests.

**Comment by taylorvnoj** [comment]
@ZakaryH 18e36292590238ba04b4f24a46c0c02dc9bcbef4 onClose is not called at Select All, but are visually selected

**Comment by ZakaryH** [comment]
got one more for ya! given how `select all` works, where it operates on the things that match your filter - if you have no results this does nothing

image

vs the clear in that same scenario, it will do something

image

probably not the worst thing ever but I'm not a huge fan of rendering things that we know will do nothing. I think this comes back to the fact that clear is not contextual whereas select is

File: packages/components/src/Combobox/Combobox.test.tsx
Comment (ZakaryH) [reply packages/components/src/Combobox/Combobox.test.tsx:405]: my final thoughts are that if this was the only thing to change that prevented an approval, it's fine. I can live with them being in here

however if we are making some changes for the `select all` button when no results, and we'd likely make some tests for that too - then I'd for sure be in favor of moving these!

**Comment by ZakaryH** [comment]
similar, when you have no options at all - we probably shouldn't have the select all button. it just doesn't make sense here imo

image

**Comment by taylorvnoj** [comment]
@chris-at-jobber can we get your opinion on[ this comment](https://github.com/GetJobber/atlantis/pull/1570#issuecomment-1769263787) - when filtering with no results should we still see Select all?

**Comment by chris-at-jobber** [comment]
@taylorvnoj makes sense to hide it if we can!

**Comment by taylorvnoj** [comment]
@ZakaryH  looking for your feedback on [this name](https://github.com/GetJobber/atlantis/pull/1570/files#diff-8a4e1042f4421ead53add52a08e0a84331aa7b33e39019aeeaff0e706ecbfc8dR17) as well.  

At first I only had `searchTermMatches` but then realized when I had a selection, then did a search with no filter results, the Clear button also disappeared.   

Otherwise, added as many tests as I could to self document the possible flows, moved those `Select` and `Clear` tests to `ComboboxContent.test` and obviously created a new subfolder for the header.

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.tsx:17]: I don't mind the name but what I do think we can do differently here is leverage work we've already done!

in `ComboboxContent` we already have a `filteredOptions` that is doing pretty much this exact operation to figure out if we have any matches

what we could do instead is 

```
 <ComboboxContentHeader
          subjectNoun={props.subjectNoun}
          selectedCount={selectedOptions.length}
          onClearAll={() => {
            optionsSelectionHandler([]);
          }}
          onSelectAll={() => {
            optionsSelectionHandler(filteredOptions);
          }}
          hasResults={filteredOptions.length > 0}
          searchValue={searchValue}
          options={props.options}
        />
```

and then add the prop `readonly hasResults: boolean;` to our types of course

this way we can cut down on iterations, DRY up the code and yeah just take advantage of work we've already done within the component!

and then I suppose to better match the `hasSelected` a name like `hasResults` or `hasSearchResults` might fit just a tiny bit better

Comment (ZakaryH) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.tsx:17]: another little benefit from this is that we no longer have to pass in `searchValue` as a prop to this

although I think the odds of reusing this might be low, it's nice to be able to de-couple those concepts. the less we need a component to know about the others, or our specific implementation - the better!

if we thought this was going to be generalized, there's even more we could do - but for now, I think letting it be just a tad bit more naive is a nice improvement

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx]: a lil nitpick here 💅 

this isn't always going to be a `ClearButton`, so a more flexible name could help communicate that

perhaps..... `showAction`? `showButton`? something like that, that's more agnostic of the function/behavior since that is dynamic

File: docs/components/Combobox/Web.stories.tsx
Comment (ZakaryH) [inline docs/components/Combobox/Web.stories.tsx:70]: oh I just noticed this. was this intentional to make the chip example a multiselect?

File: packages/components/src/Combobox/Combobox.types.ts
Comment (ZakaryH) [inline packages/components/src/Combobox/Combobox.types.ts:127]: in the vein of keeping components simple, arguably the job of the header should be
- conditionally show some text
- conditionally show a button
- do a thing when the button is pressed

an optimization we _could_ make here is to simplify the props and move the logic out of this, and into our Combobox. so out there we'd figure out if the action should be `ClearAll` or `SelectAll` and then the `Header` takes in a single action as a prop. same thing for the label logic. then, the `Header` just gets a single `Heading` prop or something like that

since I don't reasonably expect this to be re-used, I don't mind this setup and am 100% not asking for these changes, just something to consider!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.test.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.test.tsx:1]: this is fantastic test coverage that will give us so much confidence when we have to make changes in the future!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx:174]: we have this same `describe` on line `74` in this file, let's move one of these into the other and delete whichever one is now empty

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx:194]: same thing here, on line `94` we have `When the search term matches one or more options`

let's group these up!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx
Comment (ZakaryH) [inline packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx:171]: so I see we have the `onClearAll` and `onSelectAll` defined globally-ish within the scope of this larger `describe` I'm guessing that's to avoid re-defining it each time, but one downside of that is that since it's in that outer scope it will "remember" how many times it has been called, and if we had any other tests in here that made assertions around calling it, those could fail

so imagine after this test we have the inverse assertion of 

`expect(onClearAll).not.toHaveBeenCalled()`

this would fail because the same instance of `onClearAll` was called in a previous test

I'd recommend being a bit more repetitive in this instance and making a fresh, locally scoped version of `onClearAll` and `onSelectAll` for each test to do its assertions against so that it doesn't trip someone up in the future

alternatively, if we want to avoid writing the same thing out over and over, we can always make a helper function that manages the setup, gives us a fresh set of actions and then we just pass in whatever is unique to our test's setup as args. in this case that might be a bit much, but up to you!

File: docs/components/Combobox/Web.stories.tsx
Comment (taylorvnoj) [reply docs/components/Combobox/Web.stories.tsx:70]: Not really - I can't remember why I added it but I was testing something in an earlier iteration. I'll remove this!

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxHeader.tsx]: `showAction` - thank you! Did that here: fda87b77dc2f1061717036b0909a17b0bf0c9c21

File: docs/components/Combobox/Web.stories.tsx
Comment (taylorvnoj) [reply docs/components/Combobox/Web.stories.tsx]: Done here fda87b77dc2f1061717036b0909a17b0bf0c9c21

File: packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx
Comment (taylorvnoj) [reply packages/components/src/Combobox/components/ComboboxContent/ComboboxContentHeader/ComboboxContentHeader.test.tsx]: re-organized the tests here fda87b77dc2f1061717036b0909a17b0bf0c9c21

**Review by ZakaryH — APPROVED** [review]
LGTM pending passing specs

**Pull request merged**
